Fix creation of features on layers with Z

mhugo opened the pull request:
When a layer has Z coordinates (postgis for instance), loading, and modification is possible, but creation of new features or new parts is not (new features / parts are 2D only).

This patch makes new features and parts of the correct type so that they can be commited to the underlying layer.

It also make sure the postgres provider exposes the full geometry type with Z / M.

nyalldawson commented:
Mind holding off merging this till I take a look? I've got a good handle on what's required here.

mhugo commented:
Sure, let me know.

Remi-C commented:
Great ! 
Cheers

PedroVenancio commented:
Hi Hugo
Could this fix these two tickets?
https://hub.qgis.org/issues/14006
https://hub.qgis.org/issues/14007

mhugo commented:
@PedroVenancio Hi That should fix the 14007, but not the 14006 which is related to OGR apprently

mhugo commented:
@nyalldawson Hi. Are you working on some related issues ? Any progress ? May I help ? Sorry, no urge, just to know what is the state of this PR

nyalldawson commented:
Yes, sorry - will try today

nyalldawson commented on src/gui/qgsmaptoolcapture.cpp:
```diff
@@ -322,7 +332,10 @@ int QgsMapToolCapture::nextPoint( const QgsPoint& mapPoint, QgsPoint& layerPoint
+    layerPoint = QgsPointV2( toLayerCoordinates( vlayer, mapP ) ); //transform snapped point back to layer crs
+    if ( QgsWKBTypes::hasZ( QGis::fromOldWkbType( vlayer->wkbType() ) ) )
+      layerPoint.addZValue( 0.0 );
   }
```
I think we should also add 

if ( QgsWKBTypes::hasM( QGis::fromOldWkbType( vlayer->wkbType() ) ) )
      layerPoint.addMValue( 0.0 );

here. Sure, it won't do anything until the WKB types mess is sorted, but when it is that will also fix adding geometries to layers with an m dimension

mhugo replied:
Agreed.

nyalldawson commented on src/core/qgsvectorlayereditutils.h:
```diff
@@ -92,6 +92,17 @@ class CORE_EXPORT QgsVectorLayerEditUtils
+       4 if no feature was selected,
+       5 if several features are selected,
+       6 if selected geometry not found*/
+    int addPart( const QList<QgsPointV2>& ring, QgsFeatureId featureId );
```
I think while we're adding new methods here we should also fix the return value and make them return QgsVectorLayer::EditResult values to avoid these ambiguous integer returns

mhugo replied:
You mean replacing integers with enums ? 
Why not.
I would also like exceptions here

mhugo replied:
Result codes are overlapping between addPart, addRing and splitFeatures ...
I guess returning a different result code is an API break right ?
So we would need QgsVectorLayer::AddPartResult, AddRingResult, SplitResult, etc., at least before the 3.0, right ?

nyalldawson commented on src/gui/qgsmaptoolcapture.h:
```diff
@@ -88,8 +88,49 @@ class GUI_EXPORT QgsMapToolCapture : public QgsMapToolAdvancedDigitizing
   protected:
-    int nextPoint( const QgsPoint& mapPoint, QgsPoint& layerPoint );
-    int nextPoint( const QPoint &p, QgsPoint &layerPoint, QgsPoint &mapPoint );
+    /** Converts a map point to layer coordinates
```
These changes & deprecations need to be added to sip bindings too

nyalldawson commented on src/gui/qgsmaptoolcapture.h:
```diff
@@ -88,8 +88,49 @@ class GUI_EXPORT QgsMapToolCapture : public QgsMapToolAdvancedDigitizing
+          1 if the current layer is null or not a vector layer
+          2 if the transformation failed
+    */
+    int nextPoint( const QgsPointV2& mapPoint, QgsPointV2& layerPoint );
```
Also, as per the vector layer edit comment, lets fix these return values and make them return a result enum value

nyalldawson commented on src/core/qgsdatasourceuri.cpp:
```diff
@@ -543,7 +543,7 @@ QString QgsDataSourceURI::uri( bool expandAuthConfig ) const
   {
     theUri += " type=";
-    theUri += QgsWKBTypes::displayString( QgsWKBTypes::flatType(( QgsWKBTypes::Type( mWkbType ) ) ) );
+    theUri += QgsWKBTypes::displayString( QgsWKBTypes::Type( mWkbType ) );
```
This worries me a little - looks like the decoding of type relies on QGis::fromNewWkbType, which has no handling of m dimensional types and returns an undefined value. I'm not sure what effect this could have.

mhugo replied:
True.
I am thinking of keeping a new wkb type in the class and adding two methods newWkbType() and setNewWkbType(). And replace QGis::wkbType by QgsWKBTypes::Type in QGIS 3.0 ?

Commit:
Replace some magic constants with enums

Replace integers constants used for result codes
by enums in QgsGeometry, QgsGeometryEditUtils,
QgsVectorLayer, QgsVectorLayerEditUtils and QgsMApToolCapture

mhugo commented:
@nyalldawson Code updated. I replaced some magic constants by enums. But it will have to be changed for the 3.0 (use only one enum set for all the editing operations)

vpicavet commented:
Hi, can this be merged ?
@nyalldawson any further remarks ?

mhugo commented:
@nyalldawson @vpicavet I've just merged most of the commits so that it can be tested before release. There is this last commit remaining about replacing constants by enums, which I am not totally convinced yet ...

nyalldawson commented:
@mhugo I'm not familiar with this method of extending enums. But it seems rather fragile to me - if a new value is added to one of the base enums then it will clash with one of the extended values.

Why not just use a single enum for all these edit results? I *think* (can't confirm right now) that that's how qt handles similar situations.

3nids commented:
@nyalldawson there is some overlap between return values (2 has not the same meaning across the different methods). If we use a single enum it'll break the API.

nyalldawson commented:
Ok. Let's just throw a `// TODO QGIS 3.0` in and leave it as ints for now...

nyalldawson commented:
Ok. Let's just throw a ` TODO QGIS 3.0` in and leave it as ints for now...

mhugo commented:
Added TODO comments in 103856c9

slire commented:
If I'm correct, this patch should have fixed issue 14007 ( https://hub.qgis.org/issues/14007 ). However, in 2.14 it is possible to create a new feature in PostGIS PolygonZ layer by hand (it will be saved), but impossible to paste 2D objects into it (they cannot be saved with "Column has Z dimension but geometry does not" error. This means - everything is just the same as it was described in 14007 ticket, then the patch addressed some other issues?
